tests/msc2716_test.go

```diff
@@ -0,0 +1,268 @@
+
+	asUserID := "@the-bridge-user:hs1"
+	as := deployment.Client(t, "hs1", asUserID)
+	roomID := as.CreateRoom(t, struct{}{})
+
+	userID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", userID)
+	alice.JoinRoom(t, roomID, nil)
+
+	eventA, eventB, eventC, timeAfterEventA := createMessagesInRoom(t, alice, roomID)
+
+	event1, event2, event3 := backfillMessagesAtTime(t, as, roomID, eventA, timeAfterEventA)
+
+	// eventStar
+	eventStar := alice.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Message *",
+		},
+	})
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("Backfilled messages come back in correct order", func(t *testing.T) {
+			t.Parallel()
+
+			messagesRes := alice.MustDoRaw(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, nil, "application/json", url.Values{
+				"dir":   []string{"b"},
+				"limit": []string{"100"},
+			})
+
+			expectedMessageOrder := []string{
+				eventStar, eventC, eventB, event3, event2, event1, eventA,
+			}
+
+			must.MatchResponse(t, messagesRes, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONArrayEach("chunk", func(r gjson.Result) error {
+						// Find all events in order
+						if len(r.Get("content").Get("body").Str) > 0 {
+							// Pop the next message off the expected list
+							nextEventInOrder := expectedMessageOrder[0]
+							expectedMessageOrder = expectedMessageOrder[1:]
+
+							if r.Get("event_id").Str != nextEventInOrder {
+								return fmt.Errorf("Next event found was %s but expected %s", r.Get("event_id").Str, nextEventInOrder)
+							}
+						}
+
+						return nil
+					}),
+				},
+			})
+		})
+
+		t.Run("Backfilled events with m.historical do not come down /sync", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := alice.CreateRoom(t, struct{}{})
+			eventA, _, _, timeAfterEventA := createMessagesInRoom(t, alice, roomID)
+			insertOriginServerTs := uint64(timeAfterEventA.UnixNano() / 1000000)
+
+			// If we see this message in the /sync, then something went wrong
+			event1 := sendEvent(t, alice, roomID, event{
+				Type: "m.room.message",
+				PrevEvents: []string{
+					eventA,
+				},
+				OriginServerTS: insertOriginServerTs,
+				Content: map[string]interface{}{
+					"msgtype":      "m.text",
+					"body":         "Message 1",
+					"m.historical": true,
+				},
+			})
+
+			// This is just a dummy event we search for after event1
+			eventStar := sendEvent(t, alice, roomID, event{
+				Type: "m.room.message",
+				Content: map[string]interface{}{
+					"msgtype": "m.text",
+					"body":    "Message *",
+				},
+			})
+
+			// Sync until we find the star message. If we're able to see the star message
+			// that occurs after event1 without seeing event1 in the mean-time, I think we're safe to
+			// assume it won't sync
+			alice.SyncUntil(t, "", "rooms.join."+client.GjsonEscape(roomID)+".timeline.events", func(r gjson.Result) bool {
```

Comment:
Can we just use the same `SyncUntilTimelineHas` function that w also use below? it simplifies the `key` part